lib/promscrape/scrapework.go

```diff
@@ -519,23 +523,28 @@ func (sw *scrapeWork) processDataOneShot(scrapeTimestamp, realTimestamp int64, b
 	} else {
 		wc.rows.UnmarshalWithErrLogger(bodyString, sw.logError)
 	}
+	samplesPostRelabeling := 0
 	samplesScraped := len(wc.rows.Rows)
 	scrapedSamples.Update(float64(samplesScraped))
-	wc.addRows(cfg, wc.rows.Rows, scrapeTimestamp, true)
-
-	samplesPostRelabeling := len(wc.writeRequest.Timeseries)
-	if cfg.SampleLimit > 0 && samplesPostRelabeling > cfg.SampleLimit {
+	labelLimitExceeded := !wc.tryAddRows(cfg, wc.rows.Rows, scrapeTimestamp, true) && cfg.LabelLimit > 0
+	samplesPostRelabeling = len(wc.writeRequest.Timeseries)
+	if labelLimitExceeded {
+		scrapesSkippedByLabelLimit.Inc()
+		err = fmt.Errorf("the response from %q exceeds label_limit=%d; "+
+			"either reduce the labels count for exceeding series or increase label_limit", cfg.ScrapeURL, cfg.LabelLimit)
+	} else if cfg.SampleLimit > 0 && samplesPostRelabeling > cfg.SampleLimit {
+		scrapesSkippedBySampleLimit.Inc()
+		err = fmt.Errorf("the response from %q exceeds sample_limit=%d; "+
+			"either reduce the sample count for the target or increase sample_limit", cfg.ScrapeURL, cfg.SampleLimit)
+	}
```

Comment:
this logic can be moved inside tryAddRows, as we changing its singature anyway. If we do this and tryAddRows will start returning err, the processing will become much more simpler:
```
	err := wc.tryAddRows(cfg, wc.rows.Rows, scrapeTimestamp, true)
	if err != nil {
		scrapesSkippedBySampleLimit.Inc()
		wc.writeRequest.Reset()
		up = 0
	}
```